secrets/secrets-kubernetes/main.go

```diff
@@ -0,0 +1,178 @@
+	}
+	slog.Info("Get", slog.Any("application", r.Context.Application.Name), slog.Any("key", r.Name), slog.Any("impersonate", policy.Impersonate))
+
+	kubeSecretName, kubeEntryName, err := s.Locator(r.Name)
+	if err != nil {
+		return nil, fmt.Errorf("%w: invalid secret name format", secrets.ErrInvalidRequest)
+	}
+
+	kubeClient, err := kubeClientWithImpersonation(policy.Impersonate)
+	if err != nil {
+		return nil, fmt.Errorf("%w: couldn't impersonate (%s)", secrets.ErrInvalidRequest, err.Error())
+	}
+
+	kubeSecret, err := kubeClient.Secrets(policy.Namespace).Get(ctx, kubeSecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", secrets.ErrPolicy, err.Error())
+	}
+
+	kubeEntryValue, ok := kubeSecret.Data[kubeEntryName]
+	if !ok {
+		return nil, fmt.Errorf("%w: secret not found", secrets.ErrInvalidRequest)
+	}
+
+	return &secrets.Response{
+		Secret: &secrets.SecretValue{
+			Name:         r.Name,
+			StringSecret: string(kubeEntryValue),
+			Version:      kubeSecret.ResourceVersion,
+		},
+	}, nil
+}
+
+func kubeClientWithImpersonation(role string) (clientcorev1.CoreV1Interface, error) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, nil).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if role != "" {
+		config.Impersonate.UserName = role
+	}
+
+	kubeClientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return kubeClientset.CoreV1(), nil
+}
+
+func main() {
+	var (
+		natsURL            = flag.String("nats-url", nats.DefaultURL, "Nats URL")
+		secretsBackendSeed = flag.String("backend-seed", "", "NKeys Curve Seed. Leave blank for ephemeral key")
+	)
+	flag.Parse()
+
+	slog.Info("Starting", slog.Any("nats-url", *natsURL))
+
+	s := &kubeSecretsServer{}
+
+	nc, err := nats.Connect(*natsURL)
+	if err != nil {
+		slog.Error("Couldn't setup nats client", slog.Any("error", err))
+		os.Exit(1)
+	}
+
+	errorCallback := func(_ *nats.Msg, err error) {
+		slog.Error("server error", slog.Any("error", err))
+	}
+
+	var secretsBackendKey nkeys.KeyPair
+
+	if *secretsBackendSeed != "" {
+		secretsBackendKey, err = nkeys.FromSeed([]byte(*secretsBackendSeed))
+	} else {
+		slog.Info("Creating ephemeral curve keys. DO NOT USE THIS IN PRODUCTION.")
+		secretsBackendKey, err = nkeys.CreateCurveKeys()
+	}
+	if err != nil {
+		slog.Error("Couldn't setup xkey", slog.Any("error", err))
+		os.Exit(1)
+	}
+
+	secretsServer, err := secrets.NewServer(ServiceName,
+		nc,
+		s,
+		secrets.WithKeyPair(secretsBackendKey),
+		secrets.WithErrorCallback(errorCallback),
+	)
+	if err != nil {
+		slog.Error("Couldn't setup secrets server", slog.Any("error", err))
+		os.Exit(1)
+	}
+
+	mainCtx, mainCancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM, syscall.SIGUSR1, syscall.SIGUSR2, syscall.SIGHUP)
```

Comment:
I think there's a helper method somewhere in one of the k8s packages that will set this up for you. I only mention it since we're already including apimachinery, so what's one more k8s-related dependency?

Reply:
found a few buried in test packages & controller-runtime, will keep the std implementation as they will bring a world of dependencies